File: 101-symmetric-tree/symmetric-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSymmetric(self, root: Optional[TreeNode]) -> bool:
        curr = root.left

        def invert(curr):
            if curr is None:
                return curr
            
            temp = curr.left
            curr.left = curr.right
            curr.right = temp

            invert(curr.left)
            invert(curr.right)

        invert(curr)

        #compare

        def comp(p,q):
            if p==None  and q == None:
                return True
            
            if p==None  or q == None:
                return False
            
            if p.val == q.val and comp(p.left, q.left) and comp(p.right, q.right):
                return True
            return False

        return comp(root.left, root.right)

File: 101-symmetric-tree/symmetric-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isMirror(TreeNode* p, TreeNode* q){
            if (p == nullptr && q == nullptr){
                return true;
            }

            if (p == nullptr || q == nullptr){
                return false;
            }

            if (p->val == q->val && isMirror(p->left, q->right) && isMirror(p->right, q->left)){
                return true;
            }

            return false;
        }

    bool isSymmetric(TreeNode* root) {
        return isMirror(root->left, root->right);
    }
};